hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieCompactedLogRecordScanner.java

```diff
@@ -63,69 +68,96 @@
+              String lastBlockInstantTime = lastBlocks.peek().getMetadata().get(HoodieMetadataType.INSTANT_TIME);
+              if(lastBlockInstantTime.equals(blockInstantTime)) {
+                // 2 blocks (data and delete) written in the same batch, don't merge yet
+              } else {
+                // Block with the keys listed as to be deleted
+                // This is a delete block, so lets merge any records from previous data block
+                merge(records, lastBlocks);
+              }
+              // store deletes so can be rolled back
+              lastBlocks.push(r);
               break;
             case COMMAND_BLOCK:
-              log.info("Reading a command block from file " + s.getPath());
+              log.info("Reading a command block from file " + logFile.getPath());
               // This is a command block - take appropriate action based on the command
               HoodieCommandBlock commandBlock = (HoodieCommandBlock) r;
-              if (commandBlock.getType() == HoodieCommandBlockTypeEnum.ROLLBACK_PREVIOUS_BLOCK) {
-                log.info("Rolling back the last data block read in " + s.getPath());
-                // rollback the last read data block
-                recordsFromLastBlock.clear();
+              String targetInstantForCommandBlock = r.getMetadata().get(HoodieMetadataType.TARGET_INSTANT_TIME);
+              switch (commandBlock.getType()) { // there can be different types of command blocks
+                case ROLLBACK_PREVIOUS_BLOCK:
+                  // Rollback the last read log block
+                  // Get commit time from last record block, compare with targetCommitTime, rollback only if equal,
+                  // this is required in scenarios of invalid/extra rollback blocks written due to failures during
+                  // the rollback operation itself
+                  HoodieLogBlock lastBlock = lastBlocks.peek();
+                  if (lastBlock != null &&
+                          targetInstantForCommandBlock.contentEquals(lastBlock.getMetadata().get(HoodieMetadataType.INSTANT_TIME))) {
+                    log.info("Rolling back the last log block read in " + logFile.getPath());
+                    lastBlocks.pop();
+                  } else {
+                    log.warn("Invalid or extra rollback command block in " + logFile.getPath());
+                  }
+                  break;
               }
               break;
             case CORRUPT_BLOCK:
-              log.info("Found a corrupt block in " + s.getPath());
+              log.info("Found a corrupt block in " + logFile.getPath());
               // If there is a corrupt block - we will assume that this was the next data block
               // so merge the last block records (TODO - handle when the corrupted block was a tombstone written partially?)
-              merge(records, recordsFromLastBlock);
-              recordsFromLastBlock.clear();
+              merge(records, lastBlocks);
```

Comment:
do we need to merge on the corrupt block? what if a rollback comes after this?

Reply:
Yeah, good point, the reason was to overlook the corrupt block since there wasn't a good way to handle it, I've put some more logic around it now.